arwen/contexts/storage.go

```diff
@@ -238,88 +238,139 @@ func (context *storageContext) SetStorage(key []byte, value []byte) (arwen.Stora
 
 	metering := context.host.Metering()
 
-	extraBytes := len(key) - arwen.AddressLen
-	extraKeyLenGas := uint64(0)
-	if extraBytes > 0 {
-		extraKeyLenGas = math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(extraBytes))
-	}
-
-	var zero []byte
-	strKey := string(key)
 	length := len(value)
 
-	var oldValue []byte
 	storageUpdates := context.GetStorageUpdates(context.address)
-	usedCache := true
-	if update, ok := storageUpdates[strKey]; !ok {
-		// if it's not in storageUpdates, GetStorageUnmetered() will use blockchain hook for sure
-		oldValue, _ = context.GetStorageUnmetered(key)
-		storageUpdates[strKey] = &vmcommon.StorageUpdate{
-			Offset: key,
-			Data:   oldValue,
-		}
-		usedCache = false
-	} else {
-		oldValue = update.Data
+	oldValue, usedCache := context.getOldValue(storageUpdates, key)
+
+	gasForKey := context.computeGasForKey(key, usedCache)
+	metering.UseGas(gasForKey)
+
+	if bytes.Equal(oldValue, value) {
+		return context.storageUnchanged(length, usedCache)
 	}
 
-	if !usedCache || !context.flagUseDifferentGasCostForReadingCachedStorage.IsSet() {
-		metering.UseGas(extraKeyLenGas)
+	context.changeStorageUpdate(key, value, storageUpdates)
+
+	var zero []byte
+	if bytes.Equal(oldValue, zero) {
```

Comment:
I think len(oldValue) == 0 is more lightweight, but not sure

Reply:
+1 -> use len oldValue == 0